Client/core/Graphics/CRenderItem.EffectTemplate.cpp

```diff
@@ -214,6 +216,13 @@ void CEffectTemplate::CreateUnderlyingData(const SString& strFile, const SString
     macroList.back().Name = "IS_DEPTHBUFFER_RAWZ";
     macroList.back().Definition = bUsesRAWZ ? "1" : "0";
 
+    for (const auto& entry : macros)
+    {
+        macroList.push_back(D3DXMACRO());
+        macroList.back().Name = entry.first.c_str();
+        macroList.back().Definition = entry.second.c_str();
+    }
+
```

Comment:
In theory you can use `emplace_back` here:
```cpp
typedef struct D3DXMACRO {
  LPCSTR Name;
  LPCSTR Definition;
} D3DXMACRO, *LPD3DXMACRO;
```
It would be: `macroList.emplace_back(entry.first.c_str(), entry.second.c_str());`
Or a more readable way:
```cpp
for (const auto& [name, definition] : macros)
   macroList.emplace_back(name.c_str(), definition.c_str());
```
Should work just fine.

Reply:
D3DXMACRO have no constructor for two arguments. emplace_back uses std::allocator_traits so there is no way to pass the arguments this way. Are you sure MTA supporting C++17 currently? Structured binding was introduced since C++17.

Reply:
> MTA supporting C++17 currently? Structured binding was introduced since C++17.

Yes!

https://github.com/multitheftauto/mtasa-blue/blob/6d662f3f8f49a07c351be842ae27148e3cdbb407/premake5.lua#L31

But not in `game_sa` and `multiplayer_sa`:

https://github.com/multitheftauto/mtasa-blue/blob/6d662f3f8f49a07c351be842ae27148e3cdbb407/Client/game_sa/premake5.lua#L7-L9 https://github.com/multitheftauto/mtasa-blue/blob/3ab4d05e8d62278d81dcab76bd1dc89cb383e10a/Client/multiplayer_sa/premake5.lua#L7-L13